Improve Data class

carlosggarcia opened the pull request:
Improved Data class:

- [x] Avoid redundancy (cleaner code)
- [x] Improve declaration of cls to compute: npz file, all by default and none by default.
- [x] Tests

carlosggarcia commented:
@damonge  ready to review. I was thinking of changing some variable names (e.g `get_tracers_used` to `get_tracers`; `get_cl_trs_names` to `get_cl_tracers`, etc.) What do you think?

carlosggarcia commented:
@damonge now it works. So feel free to review when you can. Also, check my last commit. I think it makes sense the change I've done to the test, but worth double checking.

damonge commented on xcell/cls/data.py:
```diff
@@ -40,9 +41,149 @@ def __init__(self, data_path='', data={}, override=False,
+            default = conf
+        elif isinstance(conf, list):
+            combs = conf
+            default = self.default_clcov_from_data
```
this should be "all",  I think

damonge replied:
Not sure I understood, but sounds like you do

damonge commented on xcell/cls/data.py:
```diff
@@ -40,9 +41,149 @@ def __init__(self, data_path='', data={}, override=False,
         os.makedirs(self.data['output'], exist_ok=True)
         self._check_yml_in_outdir(override, ignore_existing_yml)
-        self.tr_matrix = None
+        self.tracers = {'wsp': None, 'no_wsp': None}
```
give this a different name

carlosggarcia replied:
I was not very imaginative so I couldn't come up with anything better. I wrote a few comments though.

damonge commented on xcell/cls/data.py:
```diff
@@ -88,24 +229,25 @@ def read_data(self, data_path):
-                tracers.append(tr2)
+        lab = ['no_wsp', 'wsp'][wsp]
+        if self.tracers[lab] is None:
+            tr_names = self._get_tracers_defined()
```
Comment: first get all tracers in yaml file

carlosggarcia replied:
Done

damonge commented on xcell/cls/data.py:
```diff
@@ -88,24 +229,25 @@ def read_data(self, data_path):
+        lab = ['no_wsp', 'wsp'][wsp]
+        if self.tracers[lab] is None:
+            tr_names = self._get_tracers_defined()
+            tmat = self.get_tracer_matrix()
```
comment: get matrix that tells you which cls you're computing

carlosggarcia replied:
This is done, I forgot to comment here

damonge commented on xcell/cls/data.py:
```diff
@@ -88,24 +229,25 @@ def read_data(self, data_path):
-            tr_nn = self.get_tracer_bare_name(tr)
-            if tr_nn in tracers:
-                tracers_for_cl.append(tr)
+            # Saving only tr1 to keep order
```
comment: `tracers` now will contain only the tracers for which we actually compute C_ells

damonge replied:
I don't think you added this comment

damonge commented on xcell/cls/data.py:
```diff
@@ -88,24 +229,25 @@ def read_data(self, data_path):
+                        tracers.append(tr1)
+                        break
+            if wsp:
+                tracers = self._filter_tracers_wsp(tracers)
```
comment: we only keep those that have different masks

carlosggarcia replied:
Done

damonge commented on xcell/cls/data.py:
```diff
@@ -184,52 +275,34 @@ def get_tracers_bare_name_pair(self, tr1, tr2, connector='-'):
-        return cl_tracers
-
     def get_cl_trs_names(self, wsp=False):
         lab = ['no_wsp', 'wsp'][wsp]
```
add a short docstring just saying what this is meant to do. Doesn't have to be long.

carlosggarcia replied:
Done

damonge commented on xcell/cls/data.py:
```diff
@@ -184,52 +275,34 @@ def get_tracers_bare_name_pair(self, tr1, tr2, connector='-'):
+        return self.cl_tracers[lab]
 
     def get_cov_trs_names(self, wsp=False):
         lab = ['no_wsp', 'wsp'][wsp]
```
add docstring as for the previous method

carlosggarcia replied:
Done

damonge reviewed (changes requested):
Haven't looked at the tests, but will do so next

carlosggarcia commented:
@damonge pinging you in case you didn't see the changes!

carlosggarcia commented:
@damonge quick summary here too. I think it only remains that you have a look at how I addressed your comments and, if you're happy with them, accept and merge

damonge reviewed (commented):
Just one thing

carlosggarcia commented:
Not sure I understand. The comment is there, isn't it?

carlosggarcia commented:
@damonge done.
